src/main/ngapp/accio/src/app/img-gallery/img-gallery.component.html

```diff
@@ -23,10 +23,74 @@
   <button type="submit" mat-button  id="upload-button">Upload</button>
 </form>
 
-<div id="gallery" *ngIf="displayImages">
+<br>
+<br>
+<br>
+<br>
+<br>
+<br>
+<br>
+<br>
+<br>
+
+<!--Filter Bar-->
+<h1>Search by:</h1>
+<div class="container">
+  <mat-form-field hintLabel="Clear to remove this filter" appearance="fill">
+    <mat-label>Image Name</mat-label>
+    <input matInput [(ngModel)]="imgName" placeholder="Ex. MyImage">
+  </mat-form-field>
+
+  <mat-form-field hintLabel="Clear to remove this filter" appearance="fill">
+    <mat-label>Mask Name</mat-label>
+    <input matInput [(ngModel)]="maskName" placeholder="Ex. MyMask">
+  </mat-form-field>
+
+  <mat-form-field hintLabel="Clear to remove this filter" appearance="fill">
+    <mat-label>Filter Tag</mat-label>
+    <input matInput [(ngModel)]="tag" placeholder="Ex. river">
+  </mat-form-field>
+
+  <mat-form-field appearance="fill">
+    <mat-label>Sort Images</mat-label>
+    <mat-select [(ngModel)]="sortImg">
+      <mat-option value="dsc">newest to oldest</mat-option>
+      <mat-option value="asc">oldest to newest</mat-option>
+    </mat-select>
+    <mat-hint align="end">Default: newest</mat-hint>
+  </mat-form-field>
+
+  <mat-form-field appearance="fill">
+    <mat-label>Show masks for Image: {{ imgName }}</mat-label>
+    <mat-select [(ngModel)]="withMasks">
+      <mat-option value="false">Don't show masks</mat-option>
+      <mat-option value="true">Show masks for: {{ imgName }}</mat-option>
+    </mat-select>
+    <mat-hint align="end">Default: no masks</mat-hint>
+  </mat-form-field>
+
+  <button mat-raised-button (click)="getImages()">Update filter(s)</button>
+</div>
+<!--Filter Bar-->
+
+<!--Fetched Images-->
+<div class="gallery">
   <div *ngFor="let image of imageArray" class="gallery-item">
     <a [routerLink]="['/editor', projectId, image['name'], image['url']]">
       <img class="gallery-img" [src]="image['url']">
+      <p>{{ image['name'] }}</p>
     </a>
   </div>
-</div>
\ No newline at end of file
+</div>
+
+<div class="gallery">
+  <ng-container *ngFor="let image of imageArray">
+    <div *ngFor="let mask of imageArray[0]['masks']" class="gallery-item">
```

Comment:
imageArray only contains one image in this case right? Could you say "*ngFor="let mask of image['masks']"

Reply:
No, because the image does not exist until the filter has been submitted by the user(has not been fetched yet). Therefore, when I have an ngFor that tries to access image['masks'] I get undefined errors. That is why I need to reference image as an element of imageArray first. Basically, if there is no image yet, my current code will not execute on any element, whereas if I changed to your suggestion, it would initially try to run on an undefined attribute.

Reply:
That should be fine